File: Wrappers/NiteControls.c/SelectableSlider2D.cpp
/*****************************************************************************
*                                                                            *
*  NiTE Controls 1.x Alpha                                                   *
*  Copyright (C) 2013 PrimeSense Ltd.                                        *
*                                                                            *
*  This file is part of NiTE Controls Lab.                                   *
*                                                                            *
*  Licensed under the Apache License, Version 2.0 (the "License");           *
*  you may not use this file except in compliance with the License.          *
*  You may obtain a copy of the License at                                   *
*                                                                            *
*      http://www.apache.org/licenses/LICENSE-2.0                            *
*                                                                            *
*  Unless required by applicable law or agreed to in writing, software       *
*  distributed under the License is distributed on an "AS IS" BASIS,         *
*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  *
*  See the License for the specific language governing permissions and       *
*  limitations under the License.                                            *
*                                                                            *
*****************************************************************************/
#include "ObjectAllocation.h"
#include "XnHash.h"
#include "PointControl.h"
#include "XnVSelectableSlider2D.h"

XN_DECLARE_DEFAULT_HASH(XnVObjectID, XnVSelectableSlider2D*, SelectableSlider2DsHash);
SelectableSlider2DsHash g_SelectableSlider2Ds;

XnVObjectID xnvSelectableSlider2D_Create(XnInt32 nXItems, XnInt32 nYItems,
													   XnFloat fSliderSizeX, XnFloat fSliderSizeY,
													   const XnChar* strName)
{
	XnVSelectableSlider2D* pSS2 = XN_NEW(XnVSelectableSlider2D, nXItems, nYItems, fSliderSizeX, fSliderSizeY, strName);
	XnVObjectID id = pSS2;//AllocateID();
	g_SelectableSlider2Ds.Set(id, pSS2);
	xnvPointControl_AddToHash(id, pSS2);

	return id;
}

XnStatus xnvSelectableSlider2D_Destroy(XnVObjectID id)
{
	xnvPointControl_RemoveFromHash(id);

	XnVSelectableSlider2D* pSS2 = NULL;
	XnStatus rc = g_SelectableSlider2Ds.Get(id, pSS2);
	if (rc != XN_STATUS_OK)
	{
		return XN_STATUS_NITE_UNEXPECTED_TYPE;
	}
	g_SelectableSlider2Ds.Remove(id);

	XN_DELETE(pSS2);
	return XN_STATUS_OK;
}
XnStatus xnvSelectableSlider2D_RegisterItemHover(XnVObjectID id,
																	   ItemHover2DCB CB,
																	   void* pCookie, XnCallbackHandle* pHandle)
{
	XN_VALIDATE_OUTPUT_PTR(pHandle);
	XnVSelectableSlider2D* pSS2 = NULL;
	XnStatus rc = g_SelectableSlider2Ds.Get(id, pSS2);
	if (rc != XN_STATUS_OK)
	{
		return XN_STATUS_NITE_UNEXPECTED_TYPE;
	}
	*pHandle = pSS2->RegisterItemHover(pCookie, CB);
	return XN_STATUS_OK;
}
XnStatus xnvSelectableSlider2D_RegisterValueChange(XnVObjectID id,
																		 ValueChange2DCB CB,
																		 void* pCookie, XnCallbackHandle* pHandle)
{
	XN_VALIDATE_OUTPUT_PTR(pHandle);
	XnVSelectableSlider2D* pSS2 = NULL;
	XnStatus rc = g_SelectableSlider2Ds.Get(id, pSS2);
	if (rc != XN_STATUS_OK)
	{
		return XN_STATUS_NITE_UNEXPECTED_TYPE;
	}
	*pHandle = pSS2->RegisterValueChange(pCookie, CB);
	return XN_STATUS_OK;
}
XnStatus xnvSelectableSlider2D_RegisterItemSelect(XnVObjectID id,
																		ItemSelect2DCB CB,
																		void* pCookie, XnCallbackHandle* pHandle)
{
	XN_VALIDATE_OUTPUT_PTR(pHandle);
	XnVSelectableSlider2D* pSS2 = NULL;
	XnStatus rc = g_SelectableSlider2Ds.Get(id, pSS2);
	if (rc != XN_STATUS_OK)
	{
		return XN_STATUS_NITE_UNEXPECTED_TYPE;
	}
	*pHandle = pSS2->RegisterItemSelect(pCookie, CB);
	return XN_STATUS_OK;
}
XnStatus xnvSelectableSlider2D_RegisterOffAxisMovement(XnVObjectID id,
																			 OffAxisMovement2DCB CB,
																			 void* pCookie, XnCallbackHandle* pHandle)
{
	XN_VALIDATE_OUTPUT_PTR(pHandle);
	XnVSelectableSlider2D* pSS2 = NULL;
	XnStatus rc = g_SelectableSlider2Ds.Get(id, pSS2);
	if (rc != XN_STATUS_OK)
	{
		return XN_STATUS_NITE_UNEXPECTED_TYPE;
	}
	*pHandle = pSS2->RegisterOffAxisMovement(pCookie, CB);
	return XN_STATUS_OK;
}
XnStatus xnvSelectableSlider2D_UnregisterItemHover(XnVObjectID id, XnCallbackHandle hCB)
{
	XnVSelectableSlider2D* pSS2 = NULL;
	XnStatus rc = g_SelectableSlider2Ds.Get(id, pSS2);
	if (rc != XN_STATUS_OK)
	{
		return XN_STATUS_NITE_UNEXPECTED_TYPE;
	}
	pSS2->UnregisterItemHover(hCB);
	return XN_STATUS_OK;
}
XnStatus xnvSelectableSlider2D_UnregisterValueChange(XnVObjectID id, XnCallbackHandle hCB)
{
	XnVSelectableSlider2D* pSS2 = NULL;
	XnStatus rc = g_SelectableSlider2Ds.Get(id, pSS2);
	if (rc != XN_STATUS_OK)
	{
		return XN_STATUS_NITE_UNEXPECTED_TYPE;
	}
	pSS2->UnregisterValueChange(hCB);
	return XN_STATUS_OK;
}
XnStatus xnvSelectableSlider2D_UnregisterItemSelect(XnVObjectID id, XnCallbackHandle hCB)
{
	XnVSelectableSlider2D* pSS2 = NULL;
	XnStatus rc = g_SelectableSlider2Ds.Get(id, pSS2);
	if (rc != XN_STATUS_OK)
	{
		return XN_STATUS_NITE_UNEXPECTED_TYPE;
	}
	pSS2->UnregisterItemSelect(hCB);
	return XN_STATUS_OK;
}
XnStatus xnvSelectableSlider2D_UnregisterOffAxisMovement(XnVObjectID id,
																   XnCallbackHandle hCB)
{
	XnVSelectableSlider2D* pSS2 = NULL;
	XnStatus rc = g_SelectableSlider2Ds.Get(id, pSS2);
	if (rc != XN_STATUS_OK)
	{
		return XN_STATUS_NITE_UNEXPECTED_TYPE;
	}
	pSS2->UnregisterOffAxisMovement(hCB);
	return XN_STATUS_OK;
}
XnStatus xnvSelectableSlider2D_Reposition(XnVObjectID id, const XnPoint3D* pCenter)
{
	XnVSelectableSlider2D* pSS2 = NULL;
	XnStatus rc = g_SelectableSlider2Ds.Get(id, pSS2);
	if (rc != XN_STATUS_OK)
	{
		return XN_STATUS_NITE_UNEXPECTED_TYPE;
	}
	pSS2->Reposition(*pCenter);
	return XN_STATUS_OK;
}
XnStatus xnvSelectableSlider2D_GetCenter(XnVObjectID id, XnPoint3D* pCenter)
{
	XnVSelectableSlider2D* pSS2 = NULL;
	XnStatus rc = g_SelectableSlider2Ds.Get(id, pSS2);
	if (rc != XN_STATUS_OK)
	{
		return XN_STATUS_NITE_UNEXPECTED_TYPE;
	}
	pSS2->GetCenter(*pCenter);
	return XN_STATUS_OK;
}
XnStatus xnvSelectableSlider2D_GetSliderSize(XnVObjectID id, XnFloat* pfSliderXSize,
													   XnFloat* pfSliderYSize)
{
	XnVSelectableSlider2D* pSS2 = NULL;
	XnStatus rc = g_SelectableSlider2Ds.Get(id, pSS2);
	if (rc != XN_STATUS_OK)
	{
		return XN_STATUS_NITE_UNEXPECTED_TYPE;
	}
	pSS2->GetSliderSize(*pfSliderXSize, *pfSliderYSize);
	return XN_STATUS_OK;
}
XnStatus xnvSelectableSlider2D_SetSliderSize(XnVObjectID id, XnFloat fSliderXSize,
													   XnFloat fSliderYSize)
{
	XnVSelectableSlider2D* pSS2 = NULL;
	XnStatus rc = g_SelectableSlider2Ds.Get(id, pSS2);
	if (rc != XN_STATUS_OK)
	{
		return XN_STATUS_NITE_UNEXPECTED_TYPE;
	}
	pSS2->SetSliderSize(fSliderXSize, fSliderYSize);
	return XN_STATUS_OK;
}
XnStatus xnvSelectableSlider2D_GetItemXCount(XnVObjectID id, XnUInt32* pnItems)
{
	XN_VALIDATE_OUTPUT_PTR(pnItems);
	XnVSelectableSlider2D* pSS2 = NULL;
	XnStatus rc = g_SelectableSlider2Ds.Get(id, pSS2);
	if (rc != XN_STATUS_OK)
	{
		return XN_STATUS_NITE_UNEXPECTED_TYPE;
	}
	*pnItems = pSS2->GetItemXCount();
	return XN_STATUS_OK;
}
XnStatus xnvSelectableSlider2D_GetItemYCount(XnVObjectID id, XnUInt32* pnItems)
{
	XN_VALIDATE_OUTPUT_PTR(pnItems);
	XnVSelectableSlider2D* pSS2 = NULL;
	XnStatus rc = g_SelectableSlider2Ds.Get(id, pSS2);
	if (rc != XN_STATUS_OK)
	{
		return XN_STATUS_NITE_UNEXPECTED_TYPE;
	}
	*pnItems = pSS2->GetItemYCount();
	return XN_STATUS_OK;
}
XnStatus xnvSelectableSlider2D_SetItemCount(XnVObjectID id, XnUInt32 nItemXCount,
													  XnUInt32 nItemYCount)
{
	XnVSelectableSlider2D* pSS2 = NULL;
	XnStatus rc = g_SelectableSlider2Ds.Get(id, pSS2);
	if (rc != XN_STATUS_OK)
	{
		return XN_STATUS_NITE_UNEXPECTED_TYPE;
	}
	pSS2->SetItemCount(nItemXCount, nItemYCount);
	return XN_STATUS_OK;
}
XnStatus xnvSelectableSlider2D_SetItemXCount(XnVObjectID id, XnUInt32 nItemXCount)
{
	XnVSelectableSlider2D* pSS2 = NULL;
	XnStatus rc = g_SelectableSlider2Ds.Get(id, pSS2);
	if (rc != XN_STATUS_OK)
	{
		return XN_STATUS_NITE_UNEXPECTED_TYPE;
	}
	pSS2->SetItemXCount(nItemXCount);
	return XN_STATUS_OK;
}
XnStatus xnvSelectableSlider2D_SetItemYCount(XnVObjectID id, XnUInt32 nItemYCount)
{
	XnVSelectableSlider2D* pSS2 = NULL;
	XnStatus rc = g_SelectableSlider2Ds.Get(id, pSS2);
	if (rc != XN_STATUS_OK)
	{
		return XN_STATUS_NITE_UNEXPECTED_TYPE;
	}
	pSS2->SetItemYCount(nItemYCount);
	return XN_STATUS_OK;
}
XnStatus xnvSelectableSlider2D_SetValueChangeOnOffAxis(XnVObjectID id, XnBool bReport)
{
	XnVSelectableSlider2D* pSS2 = NULL;
	XnStatus rc = g_SelectableSlider2Ds.Get(id, pSS2);
	if (rc != XN_STATUS_OK)
	{
		return XN_STATUS_NITE_UNEXPECTED_TYPE;
	}
	pSS2->SetValueChangeOnOffAxis(bReport);
	return XN_STATUS_OK;
}
XnStatus xnvSelectableSlider2D_GetValueChangeOnOffAxis(XnVObjectID id, XnBool* pbReport)
{
	XN_VALIDATE_OUTPUT_PTR(pbReport);
	XnVSelectableSlider2D* pSS2 = NULL;
	XnStatus rc = g_SelectableSlider2Ds.Get(id, pSS2);
	if (rc != XN_STATUS_OK)
	{
		return XN_STATUS_NITE_UNEXPECTED_TYPE;
	}
	*pbReport = pSS2->GetValueChangeOnOffAxis();
	return XN_STATUS_OK;
}

File: Wrappers/java/NiteControls.java/src/com/primesense/NiteControls/SelectableSlider2D.java
/*****************************************************************************
*                                                                            *
*  NiTE Controls 1.x Alpha                                                   *
*  Copyright (C) 2013 PrimeSense Ltd.                                        *
*                                                                            *
*  This file is part of NiTE Controls Lab.                                   *
*                                                                            *
*  Licensed under the Apache License, Version 2.0 (the "License");           *
*  you may not use this file except in compliance with the License.          *
*  You may obtain a copy of the License at                                   *
*                                                                            *
*      http://www.apache.org/licenses/LICENSE-2.0                            *
*                                                                            *
*  Unless required by applicable law or agreed to in writing, software       *
*  distributed under the License is distributed on an "AS IS" BASIS,         *
*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  *
*  See the License for the specific language governing permissions and       *
*  limitations under the License.                                            *
*                                                                            *
*****************************************************************************/
package com.primesense.NiteControls;
import org.openni.IObservable;
import org.openni.Point3D;
import org.openni.Observable;
import org.openni.OutArg;
import org.openni.StatusException;

public class SelectableSlider2D extends PointControl
{
	public SelectableSlider2D(int itemsX, int itemsY, float sliderSizeX, float sliderSizeY, String name) throws org.openni.GeneralException
	{
		super(NativeMethods.xnvSelectableSlider2D_Create(itemsX, itemsY, sliderSizeX, sliderSizeY, name));

		itemHoverEvent = new Observable<Index2EventArgs>()
		{
			@Override
			protected int registerNative(OutArg<Long> hCallback) 
			{
				return NativeMethods.xnvSelectableSlider2D_RegisterItemHover(toNative(), this, "callback", hCallback);
			}

			@Override
			protected void unregisterNative(long hCallback) 
			{
				NativeMethods.xnvSelectableSlider2D_UnregisterItemHover(toNative(), hCallback);
			}
			
			@SuppressWarnings("unused")
			public void callback(int idX, int idY)
			{
				notify(new Index2EventArgs(idX, idY));
			}
		};

		itemSelectEvent = new Observable<Index2DirectionEventArgs>()
		{
			@Override
			protected int registerNative(OutArg<Long> hCallback) 
			{
				return NativeMethods.xnvSelectableSlider2D_RegisterItemSelect(toNative(), this, "callback", hCallback);
			}

			@Override
			protected void unregisterNative(long hCallback) 
			{
				NativeMethods.xnvSelectableSlider2D_UnregisterItemSelect(toNative(), hCallback);
			}
			
			@SuppressWarnings("unused")
			public void callback(int idX, int idY, int dir)
			{
				notify(new Index2DirectionEventArgs(idX, idY, Direction.fromNative(dir)));
			}
		};
		
		valueChangeEvent = new Observable<Value2EventArgs>()
		{
			@Override
			protected int registerNative(OutArg<Long> hCallback) 
			{
				return NativeMethods.xnvSelectableSlider2D_RegisterValueChange(toNative(), this, "callback", hCallback);
			}

			@Override
			protected void unregisterNative(long hCallback) 
			{
				NativeMethods.xnvSelectableSlider2D_UnregisterValueChange(toNative(), hCallback);
			}
			
			@SuppressWarnings("unused")
			public void callback(float valueX, float valueY)
			{
				notify(new Value2EventArgs(valueX, valueY));
			}
		};
		offAxisEvent = new Observable<DirectionEventArgs>()
		{
			@Override
			protected int registerNative(OutArg<Long> hCallback) 
			{
				return NativeMethods.xnvSelectableSlider2D_RegisterOffAxisMovement(toNative(), this, "callback", hCallback);
			}

			@Override
			protected void unregisterNative(long hCallback) 
			{
				NativeMethods.xnvSelectableSlider2D_UnregisterOffAxisMovement(toNative(), hCallback);
			}
			
			@SuppressWarnings("unused")
			public void callback(int dir)
			{
				notify(new DirectionEventArgs(Direction.fromNative(dir)));
			}
		};

	}
	public SelectableSlider2D(int itemsX, int itemsY, String name) throws org.openni.GeneralException
	{
		this(itemsX, itemsY, 450, 450, name);
	}
	public SelectableSlider2D(int itemsX, int itemsY) throws org.openni.GeneralException
	{
		this(itemsX, itemsY, "SelectableSlider2D");
	}

	protected void freeObject()
	{
		super.freeObject();
		NativeMethods.xnvSelectableSlider2D_Destroy(toNative());
	}

	public void reposition(Point3D center) throws org.openni.StatusException
	{
		int status = NativeMethods.xnvSelectableSlider2D_Reposition(toNative(), center.getX(), center.getY(), center.getZ());
		org.openni.WrapperUtils.throwOnError(status);
	}
	public Point3D getCenter() throws org.openni.StatusException
	{
		OutArg<Point3D> point = new OutArg<Point3D>();
		int status = NativeMethods.xnvSelectableSlider2D_GetCenter(toNative(), point);
		org.openni.WrapperUtils.throwOnError(status);
		return point.value;
	}

	// ItemXCount
	public void setItemXCount(int items) throws StatusException
	{
		int status = NativeMethods.xnvSelectableSlider2D_SetItemXCount(toNative(), items);
		org.openni.WrapperUtils.throwOnError(status);
	}
	public int getItemXCount() throws StatusException
	{
		OutArg<Integer> items = new OutArg<Integer>();
		int status = NativeMethods.xnvSelectableSlider2D_GetItemXCount(toNative(), items);
		org.openni.WrapperUtils.throwOnError(status);
		return items.value;
	}
	// ItemYCount
	public void setItemYCount(int items) throws StatusException
	{
		int status = NativeMethods.xnvSelectableSlider2D_SetItemYCount(toNative(), items);
		org.openni.WrapperUtils.throwOnError(status);
	}
	public int getItemYCount() throws StatusException
	{
		OutArg<Integer> items = new OutArg<Integer>();
		int status = NativeMethods.xnvSelectableSlider2D_GetItemYCount(toNative(), items);
		org.openni.WrapperUtils.throwOnError(status);
		return items.value;
	}
	// SliderXSize
	public float getSliderXSize() throws StatusException
	{
		OutArg<Float> slider = new OutArg<Float>();
		int status = NativeMethods.xnvSelectableSlider2D_GetSliderXSize(toNative(), slider);
		org.openni.WrapperUtils.throwOnError(status);
		return slider.value;
	}
	// SliderYSize
	public float getSliderYSize() throws StatusException
	{
		OutArg<Float> slider = new OutArg<Float>();
		int status = NativeMethods.xnvSelectableSlider2D_GetSliderYSize(toNative(), slider);
		org.openni.WrapperUtils.throwOnError(status);
		return slider.value;
	}
	// SliderSize
	public void setSliderSize(float sliderXSize, float sliderYSize) throws StatusException
	{
		int status = NativeMethods.xnvSelectableSlider2D_SetSliderSize(toNative(), sliderXSize, sliderYSize);
		org.openni.WrapperUtils.throwOnError(status);
	}
	// ValueChangeOnOffAxis
	public void setValueChangeOnOffAxis(boolean report) throws StatusException
	{
		int status = NativeMethods.xnvSelectableSlider2D_SetValueChangeOnOffAxis(toNative(), report);
		org.openni.WrapperUtils.throwOnError(status);
	}
	public boolean getValueChangeOnOffAxis() throws StatusException
	{
		OutArg<Boolean> report = new OutArg<Boolean>();
		int status = NativeMethods.xnvSelectableSlider2D_GetValueChangeOnOffAxis(toNative(), report);
		org.openni.WrapperUtils.throwOnError(status);
		return report.value;
	}
	
	// Events
	public IObservable<Index2EventArgs> getItemHoverEvent()
	{
		return itemHoverEvent;
	}
	public IObservable<Index2DirectionEventArgs> getItemSelectEvent()
	{
		return itemSelectEvent;
	}
	public IObservable<Value2EventArgs> getValueChangeEvent()
	{
		return valueChangeEvent;
	}
	public IObservable<DirectionEventArgs> getOffAxisEvent()
	{
		return offAxisEvent;
	}
	
	private Observable<Index2EventArgs> itemHoverEvent;
	private Observable<Index2DirectionEventArgs> itemSelectEvent;
	private Observable<Value2EventArgs> valueChangeEvent;
	private Observable<DirectionEventArgs> offAxisEvent;
}
